keps/sig-cli/kubectl-apply-prune-enhancements.md

```diff
@@ -0,0 +1,273 @@
+- For one namespace, it queries all types of resources except [some hard coded ones](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply/apply.go#L632)
+to find objects that matches the label. The number of queries within one namespace is about the number of API resources installed in the cluster.
+- Among the matched objects, skip the ones without [last-applied-configuration annotation](https://kubernetes.io/docs/concepts/overview/object-management-kubectl/declarative-config/#how-apply-calculates-differences-and-merges-changes).
+- Among the rest objects, delete any object that is not present in the new apply set.
+- Repeat this for all visited namespaces. Overall, the number of queries is propotional to the number of API resources installed in the cluster.
+
+The limitations are mainly from three aspects:
+- Objects might be deleted unintentionally when different directories use the same labels
+- Some types are hard-coded to avoid being pruned.
+- The query number is neither constant nor linear to the number of resources being applied/pruned. 
+
+### A suggested solution
+
+#### Apply-Config Object
+To solve the first challenge, a concept of apply-config object is proposed. An apply-config object is an object that contains the GKN information for all the objects that are to be applied. The apply-config object is not generated by kubectl when applying is running. Instead, it should be generated by other tools such as kustomize, Helm. When generating resources to be consumed by `kubectl apply`, those tools are responsible to generate the associated apply-config object to utilize better pruning.
+
+An object is identified as an apply-config object when it has a special label `kubectl.kubernetes.io/apply-cfg: "true"`. The GKN information for all other objects are stored in annotations. Here is an apply-config object example.
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: name-foo
+ labels:
+   kubectl.kubernetes.io/apply-cfg: "true"
+ annotations:
+   kubectl.kubernetes.io/last-applied-objects: |
+     namespace1:group1:kind1:somename
+     namespace1:group2:kind2:secondname
+     namespace3:group3:kind3:thirdname
+```
+Note that the version is not necessary to find the object. The version of the resource can be selected by the client and available versions can be discovered dynamically
+
+The apply-config object is not limited to ConfigMap type. Any type of Object can be an apply-config object as long as it has the `apply-cfg` label. That means a custom resource object, such as an [Application](https://github.com/kubernetes-sigs/application) custom resource object, can also be used as an apply-config object when needed.
+
+The apply-config object could have an optional hash label `kubectl.kubernetes.io/apply-hash`, which is based on the content of annotation `kubectl.kubernetes.io/last-applied-objects`. This hash label is used for validation in kubectl. More details are in [How It Works](#how-it-works).
```

Comment:
This annotation is going away.

If you say it "could" have something, it's useful to say why it might be used (validation of what?). Or just defer discussion completely to the section mentioned.

Reply:
Will move it to later parts.